core/pkg/sync-store/sync_store.go

```diff
@@ -0,0 +1,287 @@
+				target,
+				key,
+			))
+		sh = &syncHandler{
+			dataSync: make(chan isync.DataSync),
+			subs: map[interface{}]storedChannels{
+				key: {
+					errChan:  errChan,
+					dataSync: dataSync,
+				},
+			},
+			mu: &sync.RWMutex{},
+		}
+		s.syncHandlers[target] = sh
+		go s.watchResource(target)
+	} else {
+		// register our sub in the map
+		s.logger.Debug(fmt.Sprintf("registering sync subscription %p", key))
+		sh.subs[key] = storedChannels{
+			errChan:  errChan,
+			dataSync: dataSync,
+		}
+
+		// access pointer + trigger resync passing the dataSync
+		if sh.syncRef != nil {
+			go func() {
+				s.logger.Debug(fmt.Sprintf("sync handler exists for target %s, triggering a resync", target))
+				if err := sh.syncRef.ReSync(ctx, dataSync); err != nil {
+					errChan <- err
+				}
+			}()
+		}
+	}
+	// defer until context close to remove the key
+	go func() {
+		<-ctx.Done()
+		s.mu.Lock()
+		if s.syncHandlers[target] != nil && s.syncHandlers[target].subs != nil {
+			s.logger.Debug(fmt.Sprintf("removing sync subscription due to context cancellation %p", key))
+			delete(s.syncHandlers[target].subs, key)
+		}
+		s.mu.Unlock()
+	}()
+}
+
+func (s *SyncStore) watchResource(target string) {
+	s.logger.Debug(fmt.Sprintf("watching resource %s", target))
+	ctx, cancel := context.WithCancel(s.ctx)
+	defer cancel()
+	sh, ok := s.syncHandlers[target]
+	if !ok {
+		s.logger.Error(fmt.Sprintf("no sync handler exists for target %s", target))
+		return
+	}
+	// this cancel is accessed by the cleanup method shutdown the listener + delete the syncHandler
+	sh.cancelFunc = cancel
+	go func() {
+		<-ctx.Done()
+		s.mu.Lock()
+		delete(s.syncHandlers, target)
+		s.mu.Unlock()
+	}()
+	// broadcast any data passed through the core channel to all subscribing channels
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case d := <-sh.dataSync:
+				sh.writeData(s.logger, d)
+			}
+		}
+	}()
+	// setup sync, if this fails an error is broadcasted, and the defer results in cleanup
+	sync, err := s.syncBuilder.SyncFromURI(target, s.logger)
+	if err != nil {
+		s.logger.Error(fmt.Sprintf("unable to build sync from URI for target %s: %s", target, err.Error()))
+		sh.writeError(s.logger, err)
+		return
+	}
+	// init sync, if this fails an error is broadcasted, and the defer results in cleanup
+	err = sync.Init(ctx)
+	if err != nil {
+		s.logger.Error(fmt.Sprintf("unable to initiate sync for target %s: %s", target, err.Error()))
+		sh.writeError(s.logger, err)
+		return
+	}
+	// sync ref is used to trigger a resync on a single channel when a new subscription is started
+	// but the associated SyncHandler already exists, i.e. this function is not run
+	sh.syncRef = sync
+	err = sync.Sync(ctx, sh.dataSync)
+	if err != nil {
+		s.logger.Error(fmt.Sprintf("error from sync for target %s: %s", target, err.Error()))
+		sh.writeError(s.logger, err)
+	}
+}
+
+func (h *syncHandler) writeError(logger *logger.Logger, err error) {
+	h.mu.RLock()
```

Comment:
Minor - A deferred unlock is better as it guarantees the unlock :) 

```suggestion
	h.mu.RLock()
	defer h.mu.RUnlock()
```